go/ql/test/library-tests/semmle/go/frameworks/Twirp/tests.ql

```diff
@@ -0,0 +1,188 @@
+import go
+import semmle.go.frameworks.Twirp
+
+class InlineTest extends LineComment {
+  string tests;
+
+  InlineTest() { tests = this.getText().regexpCapture("\\s*test:(.*)", 1) }
+
+  string getPositiveTest() {
+    result = tests.trim().splitAt(",").trim() and not result.matches("!%")
+  }
+
+  string getNegativeTest() { result = tests.trim().splitAt(",").trim() and result.matches("!%") }
+
+  predicate hasPositiveTest(string test) { test = this.getPositiveTest() }
+
+  predicate hasNegativeTest(string test) { test = this.getNegativeTest() }
+
+  predicate inNode(DataFlow::Node n) {
+    this.getLocation().getFile() = n.getFile() and
+    this.getLocation().getStartLine() = n.getStartLine()
+  }
+
+  predicate inEntity(Entity e) {
+    this.getLocation().getFile() = e.getDeclaration().getFile() and
+    this.getLocation().getStartLine() = e.getDeclaration().getLocation().getStartLine()
+  }
+
+  predicate inType(Type t) {
+    exists(TypeEntity te |
+      te.getType() = t and
+      this.getLocation().getFile() = te.getDeclaration().getFile() and
+      this.getLocation().getStartLine() = te.getDeclaration().getLocation().getStartLine()
+    )
+  }
+}
+
+query predicate passingPositiveTests(string res, string expectation, InlineTest t) {
+  res = "PASSED" and
+  t.hasPositiveTest(expectation) and
+  (
+    expectation = "handler" and
+    exists(Twirp::ServiceHandler n | t.inEntity(n))
+    or
+    expectation = "request" and
+    exists(Twirp::Request n | t.inNode(n))
+    or
+    expectation = "ssrfSink" and
+    exists(RequestForgery::Sink n | t.inNode(n))
+    or
+    expectation = "message" and
+    exists(Twirp::ProtobufMessageType n | t.inType(n))
+    or
+    expectation = "serviceInterface" and
+    exists(Twirp::ServiceInterfaceType n | t.inType(n.getNamedType()))
+    or
+    expectation = "serviceClient" and
+    exists(Twirp::ServiceClientType n | t.inType(n))
+    or
+    expectation = "serviceServer" and
+    exists(Twirp::ServiceServerType n | t.inType(n))
+    or
+    expectation = "clientConstructor" and
+    exists(Twirp::ClientConstructor n | t.inEntity(n))
+    or
+    expectation = "serverConstructor" and
+    exists(Twirp::ServerConstructor n | t.inEntity(n))
+    or
+    expectation = "ssrf" and
+    exists(RequestForgery::Configuration cfg, DataFlow::Node sink |
+      cfg.hasFlow(_, sink) and t.inNode(sink)
+    )
```

Comment:
What I meant was copy `go/ql/test/query-tests/Security/CWE-918/RequestForgery.qlref` into your test folder, so that that query is run as part of the tests, and write some test code which you'd expect to produce results with the new sources and sinks you've defined.

Reply:
hmmm, the `RequestForgery` query is not reporting the issue. Apparently neither the new source nor the new sink match the Configuration's `isSource/isSink` predicates. However, my test above works. Any ideas?

Reply:
Ok, figured out the problem, I needed to import the Twirp module from `go.qll`